autrainer/core/scripts/preprocess_script.py

```diff
@@ -118,110 +117,36 @@ def _preprocess_datasets(self) -> None:
-
-            file_count = 0
-            for file_path in chunk:
-                in_path = Path(dataset["path"], "default", file_path)
-                out_path = Path(
-                    dataset["path"],
-                    dataset["features_subdir"],
-                    file_path,
-                ).with_suffix("." + dataset["file_type"])
-                data = file_handler.load(in_path)
-                data = pipeline(data, 0)
-                output_file_handler.save(out_path, data)
-                del data
-                file_count += 1
-                if file_count % self.update_frequency == 0:
-                    with lock:
-                        pbar.update(self.update_frequency)
-            with lock:
-                pbar.update(file_count % self.update_frequency)
-
         print("Preprocessing datasets...")
         for (name, dataset), preprocess in zip(
             self.datasets.items(), self.preprocessing.values()
         ):
             print(f" - {name}")
-            # swap dataset handler with preprocessing handler
-            features_subdir = dataset["features_subdir"]
+            if preprocess is None:
+                print("No preprocessing specified. Skipping...")
+                continue
+            # set dataset file handler as output handler
+            # as dataset is already configured
+            # to work with the output of the preprocessing script
             output_file_handler = autrainer.instantiate_shorthand(
                 config=dataset["file_handler"],
                 instance_of=AbstractFileHandler,
             )
             output_file_type = dataset["file_type"]
-            # set features_subdir to None (defaults to audio)
-            # so that iteration takes place over raw data
-            dataset["features_subdir"] = None
+            
+            # override dataset file handling to work with raw audio
             dataset["file_handler"] = preprocess["file_handler"]
-            dataset["file_type"] = preprocess["file_type"]
-            if preprocess is None:
-                continue
+            # None allows dataset to work with all audio files
+            dataset["file_type"] = None
+            dataset["seed"] = 0  # ignored
+            dataset["batch_size"] = 8  # ignored
             data = autrainer.instantiate_shorthand(dataset)
-            loader = torch.utils.data.DataLoader(
-                torch.utils.data.ConcatDataset(
-                    data.train_dataset,
-                    data.dev_dataset,
-                    data.test_dataset
-                ),
-                shuffle=False,
-                num_workers=self.num_workers,
-                batch_size=1  #TODO: can we do it batched?
-            )
+            # manually disable dataset transforms
+            data.train_transform = None
+            data.dev_transform = None
+            data.test_transform = None
+            os.makedirs(data.features_subdir, exist_ok=True)
```

Comment:
We still want to enforce the location of the features subdir to be alongside the original path right (`{path}/log_mel_16k/spectrogram.npy`)?
This would be the path constructed right now (which at the same time allows us to conveniently check the matching preprocessing transform to apply).
This would lead to issues if the user has no write access to `{path}`.
The solution would be to override `audio_subdir` with an absolute path (pointing to the original dataset format, possible only with read access) as well as the dataframe properties to depend on `audio_subdir` as well.
Finally using `{path}` as the local place for feature extraction only.

Was this the intention that you had in mind?

Reply:
Ah, yes, another unforeseen complication. I'll think about it a bit, but I guess we might need to disentangle the two paths..

Reply:
75c07d8 added an optional `features_path` argument to `AbstractDataset`. Whenever it is provided, it overrides `self.path` as far as **features** are concerned (labels remain untouched).

There were some "hacks" to get this to work with `preprocess`, but I've finally been able to "test" this with EmoDB and it seems to work, both with the default `features_path=None` which stores features along with the original data, and with `features_path="features"` which uses a different path. The training also works with both.

This is a pretty big change under the hood though, which I'd like to have included in the tests. Do you have any idea how? Doesn't seem to be compatible with our `ToyDatasets` right now..

Reply:
Yeah it will be tricky to test with the `ToyDataset` and i would think these tests should be done for the base datasets where i have a small setup for mocking data already which you could reuse:
https://github.com/autrainer/autrainer/blob/2485ef339bcd90e96f4b63a0dedcb6c99358dd0c/tests/test_datasets.py#L80-L98

Reply:
Added some tests in a0254a2

I also moved the core processing of the script to a new function, which made it easier to test.

This will also be useful if we provide a "proper" API in the future, as it's an entry-point for users to do the preprocessing in `python`